py-tests/pyspark_tests.py

```diff
@@ -0,0 +1,642 @@
+    ca_path = home+'/.docker/machine/certs/ca.pem'
+    tls_config = docker.tls.TLSConfig(client_cert=(cert_path, key_path), verify=ca_path)
+    cli = docker.Client(base_url=u'tcp://192.168.99.100:2376', tls=tls_config)
+
+    #cli.pull('basho/riak-ts', stream=True)
+
+    try:
+        cli.stop('dev1')
+        cli.remove_container('dev1')
+    except Exception as e:
+        pass
+
+    riak_container = cli.create_container(image='basho/riak-ts', name='dev1', ports=[8087, 8098], \
+                                          host_config=cli.create_host_config(port_bindings={8087: 8087, 8098: 8098}))
+    cli.start('dev1')
+
+    #request.addfinalizer(lambda: cli.remove_container('dev1'))
+    #request.addfinalizer(lambda: cli.stop('dev1'))
+    return cli
+@pytest.mark.usefixtures("docker_cli")
+
+
+@pytest.fixture(scope="session")
+def spark_context(request, docker_cli):
+
+    host, pb_port, hostAndPort = get_host_and_port(docker_cli)
+    conf = (SparkConf().setMaster("local[*]").setAppName("pytest-pyspark-local-testing"))
+    conf.set('spark.riak.connection.host', hostAndPort)
+    sc = SparkContext(conf=conf)
+    sc.setLogLevel("ERROR")
+    pyspark_riak.riak_context(sc)
+    request.addfinalizer(lambda: sc.stop())
+    return sc
+@pytest.mark.usefixtures("spark_context")
+
+
+@pytest.fixture(scope="session")
+def sql_context(request, spark_context):
+
+    sqlContext = SQLContext(spark_context)
+    return sqlContext
+@pytest.mark.usefixtures("sql_context")
+
+
+@pytest.fixture(scope="session")
+def riak_client(request, docker_cli):
+
+    host, pb_port, hostAndPort = get_host_and_port(docker_cli)
+    client = riak.RiakClient(host=host, pb_port=pb_port)
+    request.addfinalizer(lambda: client.close())
+    return client
+@pytest.mark.usefixtures("riak_client")
+
+###### FUNCTIONS #######
+
+def get_host_and_port(docker_cli):
+
+    container = [x for x in docker_cli.containers() if x['Names'][0] == u'/dev1']
+    host= container[0]['NetworkSettings']['Networks']['bridge']['IPAddress']
+    pb_port = container[0]['Ports'][1]['PublicPort']
+    hostAndPort = ":".join([str(host), str(pb_port)])
+    return host, pb_port, hostAndPort
+
+@timeout_decorator.timeout(1, use_signals=False)
+def ping_with_timeout_1_second(client):
+    return client.ping()
+
+@timeout_decorator.timeout(1, use_signals=False)
+def store_obj_timeout_1_second(client, bucket, key, data):
+
+    temp_bucket = client.bucket(bucket)
+    obj = riak.RiakObject(client, temp_bucket, key)
+    obj.content_type = 'text/plain'
+    obj.data = data
+    obj.store()
+
+@timeout_decorator.timeout(1, use_signals=False)
+def read_obj_timeout_1_second(client, bucket, key):
+    return client.bucket(bucket).get(key).data
+
+def wait_until_ping(client,times=20):
+
+    pinged = False
+    i = 0
+    while not pinged and i < times:
+        try:
+            pinged = ping_with_timeout_1_second(client)
+        except:
+            pass
+        i = i + 1
+        time.sleep(1)
```

Comment:
The `sleep` is going to execute even if the ping is successful. AND this defect is in multiple locations since the retry-specific logic was not factored out.

A fix for the spurious `sleep` would be to have the statements to execute in the case of error, which include `sleep`as well as incrementing `times`, to be in the `except` block. Another fix would be to return immediately on success. For a point of comparison, see https://github.com/basho/cache_proxy/blob/develop/tests/lib/utils.py#L178